### ISSUE TITLE
Bug: `get_args` does not return the correct tuple of types

### ISSUE THREAD

**Issue by guacs** [pydantic/pydantic#7837]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

Pydantic `BaseModel` does not expose the `__args__` field for generic types. This means that `get_args` also fails to return the correct tuple of types for a generic class. The only way to get it is via the `__pydantic_generic_metadata__`. I'm marking it as a bug because the `__parameters__` field is available.

Also, `get_origin` does not work on generics either.

### Example Code

```Python
from typing import Generic, TypeVar

from pydantic import BaseModel
from typing_extensions import get_args

T = TypeVar("T")


class Foo(BaseModel, Generic[T]):
    foo: T


FooInt = Foo[int]
args = getattr(FooInt, "__args__", ())

assert args == get_args(FooInt) == (int,), args
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.4.2
        pydantic-core version: 2.10.1
          pydantic-core build: profile=release pgo=true
                 install path: ~
               python version: 3.12.0 (main, Oct  3 2023, 18:39:42) [GCC 11.4.0]
                     platform: Linux-5.15.90.1-microsoft-standard-WSL2-x86_64-with-glibc2.35
             related packages: email-validator-2.0.0.post2 pyright-1.1.331 pydantic-extra-types-2.1.0 mypy-1.6.0 typing_extensions-4.8.0
```

**Comment by sydney-runkle** [comment]
Hi @guacs,

Thanks for reporting this. As seen [here](https://github.com/pydantic/pydantic/issues/7815#issuecomment-1772821044), we currently have `pydantic._internal._generics.get_args` and `pydantic._internal._generics.get_origin` available. The comment I've linked goes into more detail about adding support for `typing.get_args` and `typing.get_origin` functions in the future 😄.

PRs are more than welcome 👍

**Comment by alexmojaki** [comment]
> Pydantic `BaseModel` does not expose the `__args__` field for generic types. This means that `get_args` also fails to return the correct tuple of types for a generic class.

It's not enough to expose `__args__`. [Here](https://github.com/python/cpython/blob/0db2517687efcf5ec0174a32398ec1564b3204f1/Lib/typing.py#L2268-L2290) is what the source code of `typing.get_args` looks like:

```python
def get_args(tp):
    if isinstance(tp, _AnnotatedAlias):
        ...
    if isinstance(tp, (_GenericAlias, GenericAlias)):
        res = tp.__args__
        ...
    if isinstance(tp, types.UnionType):
        ...
    return ()
```

So in addition to providing `__args__`, `tp` (e.g. `Foo[int]`) needs to be an instance of one of those classes. There's no magic methods or protocols that can be implemented to work around that unless `typing.get_args` changes.

So let's suppose that `isinstance(MyModelClass, _GenericAlias)` is `True`, because this has seemed desirable in other situations, and `_GenericAlias` seems the most sensible option of the classes above. This causes `isinstance(MyModelClass(...), typing.Pattern)` to raise an exception (details at the end). I mention this because `pytest` performs such a check which causes the tests to crash before they can begin, but either way it's an innocent check that shouldn't be raising an error. The reason that this happens is because of [this method in the `_SpecialGenericAlias` class](https://github.com/python/cpython/blob/main/Lib/typing.py#L1426-L1431) (`Pattern` is an instance of `_SpecialGenericAlias`, and presumably other instances of this class cause similar problems):

```python
    def __subclasscheck__(self, cls):
        ...
        if not isinstance(cls, _GenericAlias):
            # previously the execution would go here where things are fine
            return issubclass(cls, self.__origin__)
        # now we go here where the error happens
        return super().__subclasscheck__(cls)
```

Here `self` is `typing.Pattern` and `cls` is `MyModelClass`. This means that `isinstance(cls, _GenericAlias)` returning `True` (which was our goal) caused an error. Without changes to `typing.py`, this is unavoidable. And if somehow it wasn't, or if we ignored this problem, it's an example that shows how easily changing the result of `isinstance` could cause unforeseeable problems in third-party code.

@dmontagu in https://github.com/pydantic/pydantic/issues/7815#issuecomment-1772821044 you wrote:

> (Open to a PR if anyone can figure out how to make that work without breaking changes; it would require making parametrized submodels be subclasses of `typing._GenericAlias`, I have tried some things in the past but not been able to resolve all the challenges.)

which seems to imply that the challenges you ran into looked like they might be theoretically solvable. I think I've shown that they're not, does that seem like a fair conclusion?

So these are the options I see for this issue:

1. Consider reopening https://github.com/pydantic/pydantic/pull/7894. If you have specific concrete concerns about this, please share them. I can think of a couple of problems that are worth addressing but are very solvable.
2. Settle for achieving some goal other than making `typing.get_args` work, e.g. exposing and documenting some version of `pydantic._internal._generics.get_args`. However this doesn't help if the call to `get_args` is in some third party code outside of the control of users.
3. Close the issue and leave it unresolved.

Separately, it might be worth thinking about proposing changes to Python to make it more flexible/extensible. This issue alone may not be enough for that, but there's also https://github.com/pydantic/pydantic/issues/6994#issuecomment-1773295810, where monkeypatching was much less successful and the problem wasn't just making something a `_GenericAlias` but making something that _wasn't a class_. I imagine there might be other issues with similar problems.

-----

## Reproduction details

This change:

```diff
diff --git a/pydantic/_internal/_model_construction.py b/pydantic/_internal/_model_construction.py
index d7423ce8..3496e889 100644
--- a/pydantic/_internal/_model_construction.py
+++ b/pydantic/_internal/_model_construction.py
@@ -68,7 +68,10 @@ class _ModelNamespaceDict(dict):
 
 
 @dataclass_transform(kw_only_default=True, field_specifiers=(Field,))
-class ModelMetaclass(ABCMeta):
+class ModelMetaclass(ABCMeta, typing._GenericAlias, _root=True):
+    __origin__ = None
+    __args__ = ()
+
     def __new__(
         mcs,
         cls_name: str,
```

(without `_root=True` you get `TypeError: Cannot subclass special typing classes` because Python really doesn't want us to do this)

is the bare minimum required to run this script:

```python
import typing
from pydantic import BaseModel

class M(BaseModel):
    pass

assert isinstance(M, typing._GenericAlias)
isinstance(M(), typing.Pattern)  # error here
```

which gives this traceback:

```pytb
Traceback (most recent call last):
  File "/home/alex/.config/JetBrains/PyCharm2023.2/scratches/scratch_1370.py", line 11, in <module>
    isinstance(M(), typing.Pattern)
  File "/home/alex/.pyenv/versions/3.11.5/lib/python3.11/typing.py", line 1305, in __instancecheck__
    return self.__subclasscheck__(type(obj))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/alex/.pyenv/versions/3.11.5/lib/python3.11/typing.py", line 1584, in __subclasscheck__
    return super().__subclasscheck__(cls)     ### <<<----- this is in _SpecialGenericAlias
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/alex/.pyenv/versions/3.11.5/lib/python3.11/typing.py", line 1308, in __subclasscheck__
    raise TypeError("Subscripted generics cannot be used with"
TypeError: Subscripted generics cannot be used with class and instance checks
```

We actually want something like `assert isinstance(M[int], typing._GenericAlias)` (i.e. `M[int]` instead of just `M`), but to make generics even begin to work you have to make additional changes like moving `BaseModel.__class_getitem__` to `ModelMetaclass.__getitem__`. This is also necessary to get to the point where `pytest` calls `isinstance(val, Pattern)`.

Note that `isinstance` calls `Pattern.__instancecheck__`. `ModelMetaclass.__instancecheck__` is already defined but it simply isn't called at any point.

Subclassing `GenericAlias` (not `_GenericAlias`) gives:

```
    class ModelMetaclass(ABCMeta, typing.GenericAlias):
TypeError: multiple bases have instance lay-out conflict
```

Maybe there's another way to get `M[int]` to be a `GenericAlias` but I'm not going to try and it really looks like a lost cause.

**Comment by dmontagu** [comment]
The idea would not be for BaseModel to be a subclass of `_GenericAlias`, but for `BaseModel.__class_getitem__` to return something that was an instance of a subclass of `_GenericAlias` properly modified to proxy dunder attributes etc. necessary for pydantic in ways that standard `_GenericAlias` subclass instances do not.

@adriangb looked into this in some depth a while ago.

I will look a bit more deeply into your comment but like I said, I still think it should be possible, just will require some (significant) changes to what a parametrized generic model "is" and what gets returned from `BaseModel.__class_getitem__` specifically.

I'll also note that I think as long as the user-facing APIs are consistent, in principle I'd be okay making changes like `BaseModel.__class_getitem__` not actually returning a subclass, but instead a `_GenericAlias instance` (which might itself reference a created subclass). It won't surprise me if that ends up being "breaking" enough of a change that it needs to wait for v3, but like I said I think it should be possible to do sufficiently well that we think it's worth changing the `BaseModel.__class_getitem__` implementation at a point where we are okay with more significant changes.

------

I also think @adriangb had strong opinions about not wanting to monkeypatch the stdlib typing library, which I think I generally agree with. I personally don't have a super strong concrete argument, but intuitively to me it just feels fragile/invasive. I don't like when libraries I use modify anything about stdlib (or other libraries I depend on), it just makes everything more confusing while trying to debug / trace code execution into the affected functions.

**Comment by adriangb** [comment]
I agree with David. I would like us to explore using `types.GenericAlias` subclasses (i.e. behaving more like standard library generics like `list` etc., `list[int].sort` works) not only because it would fix this issue but because it has the potential to improve lots of other things and make Pydantic itself faster (the current approach of returning a fully functional class is expensive). Only if that fails would I even consider monkey patching the standard library.

That said there is precedent: typing-extensions does it all the time. But the authors of that library are also the authors of `typing.py` itself, so they really know what they're doing and keep up to date with changes quickly. Maybe this is even an argument for _not_ monkey patching: there is already a library installed pretty much everywhere that does monkey patching of the same functions , we might conflict with it.

**Comment by dmontagu** [comment]
Looking a bit more closely, I see that your error in the subclass check is because you made `ModelMetaclass` a subclass of `typing._GenericAlias`, which I don't think makes sense. You only get `typing._GenericAlias` instances from standard library generics when you actually parametrize them, and when you try to "instantiate" a `typing._GenericAlias`, it proxies to the underlying class. That is, the following runs without errors:

```python
import typing

T = typing.TypeVar("T")


class M(typing.Generic[T]):
    pass


assert isinstance(M[int], typing._GenericAlias)
assert not isinstance(M[int](), typing._GenericAlias)
assert isinstance(M[int](), M)
```

So basically, I definitely don't think there's a world where it will work to make `ModelMetaclass` inherit from `typing._GenericAlias`, but I don't think that's what should be done anyway. When I said @adriangb looked into this before, I meant that he tried making a subclass of `typing._GenericAlias` for which instances would be returned from `__class_getitem__`. I think the main challenges he ran into were with getting recursive generics to work (I'm not 100% sure about that though). But I think it's achievable, just a lot of painful work that was hard to justify at the time (we were racing to get v2 released on a reasonable schedule).

**Comment by adriangb** [comment]
Note that `types.GenericAlias` is not `typing._GenericAlias` and the former is what we should be using not the latter.

**Comment by dmontagu** [comment]
(Also, I understand that what I originally wrote — "require making parametrized submodels be subclasses of typing._GenericAlias" — was not equivalent to my corrected/clarified approach described in my more recent messages in this thread, apologies for confusion, I agree with your assessment that it can't work to make ModelMetaclass a subclass of typing._GenericAlias or whatever but, yeah I think the approach described above using instances of `types.GenericAlias` in as the return value of `__class_getitem__` may be more tractable.)

**Comment by alexmojaki** [comment]
> Looking a bit more closely, I see that your error in the subclass check is because you made ModelMetaclass a subclass of typing._GenericAlias, which I don't think makes sense.

I left that at the end as an implementation detail. The point is that `isinstance(MyModelClass, _GenericAlias)` being True seemed to imply that `isinstance(MyModelClass(...), typing.Pattern)` must raise an error, which seems like a dealbreaker. This would also be the case if `MyModelClass = MyGenericClass[int]` so it didn't matter if the implementation used `ModelMetaclass` or `__class_getitem__`.

But I do see now that there's a hole in my argument which IIUC might be what you're saying, which is that we could have this:

```python
assert isinstance(M[int], typing._GenericAlias)
assert isinstance(M[int](), M[int])
assert type(M[int]()) is not M[int]
```

i.e. `M[int]()` returns something that 'is an instance' of `M[int]` but that doesn't mean that it instantiates `M[int]` directly.

Is `type(M[int]()) is not M[int]` considered acceptable? It's definitely a change in behaviour and it looks scary, but I can see how maybe it could be OK, since we'd expect users to generally use `isinstance` or at least `type(M[int]()) == M[int]` which we could maybe handle.

But if `type(M[int]()) is M[int]` then the problem before remains regardless of implementation: `isinstance(M[int](), typing.Pattern)`calls `_BaseGenericAlias.__instancecheck__(typing.Pattern, M[int]())` which calls `typing.Pattern.__subclasscheck__(type(M[int]()))` which means `_SpecialGenericAlias.__subclasscheck__(typing.Pattern, M[int])` which fails if `isinstance(M[int], typing._GenericAlias)`.

> Note that `types.GenericAlias` is not `typing._GenericAlias` and the former is what we should be using not the latter.

Thanks, this is helpful and interesting, other conversations definitely focused on `typing._GenericAlias`. I see that `types.GenericAlias` is new in Python 3.9 and doesn't seem to have an equivalent in typing-extensions, is that a problem?

Do either of you have any partial code or other resources that would be helpful in trying to make `M[int]` an instance of `types.GenericAlias`?

FWIW the fact that `isinstance(M[int], typing._GenericAlias)` came so close to causing an unavoidable error after barely poking the surface doesn't leave me very optimistic that `isinstance(M[int], types.GenericAlias)` will be free of problems. Quoting myself:

> And if somehow it wasn't, or if we ignored this problem, it's an example that shows how easily changing the result of `isinstance` could cause unforeseeable problems in third-party code.

So even if all the tests are passing it could easily still be unsafe.

**Comment by dmontagu** [comment]
> Is `type(M[int]()) is not M[int]` considered acceptable?

Considering that's the case for all standard library generics, I would say yes it is acceptable, though I would guess it would be a breaking change for enough peoples' code that we probably(?) shouldn't make the change before pydantic v3.

> if `type(M[int]()) is M[int]` then the problem before remains

While I'm not super convinced that it's the end of the world if checking if `M[int]` is an instance of `typing.Pattern` raises an error, in general I agree that it's very likely not possible to get `assert type(M[int]()) is M[int]` to work while also having `typing.get_args` etc. work. But again, I think that's closely related to the fact that `type(M[int]()) is not M[int]` is the case for all standard library generics (as far as I'm aware).

**Comment by dmontagu** [comment]
I'll also note that with standard library generics:
```
assert isinstance(M[int](), M[int])
```
would actually raise an error, and in general I think it's not a great idea to be doing subclass checks of parametrized generics; I think there's a tendency for people to get confused by covariance, etc. and think things like `isinstance(M[int](), M[Any])` should be true. Which might be reasonable in many circumstances, but isn't the case due to implementation details in pydantic. (And the stdlib `typing` module avoids the confusion by just pre-emptively erroring.)

**Comment by alexmojaki** [comment]
> I would guess it would be a breaking change for enough peoples' code that we probably(?) shouldn't make the change before pydantic v3.

If something has to wait until v3, then I sort of expect that it would have a special 'v3' label meaning "we're still recovering from releasing v2 and don't want too spend much energy thinking about this yet". What's the general attitude on your side?

Would changing `assert isinstance(M[int](), M[int])` be acceptable even in v3? It seems like a hell of a change.

(of course all this is only relevant if similar problems appear for `types.GenericAlias`)

**Comment by alexmojaki** [comment]
> While I'm not super convinced that it's the end of the world if checking if `M[int]` is an instance of `typing.Pattern` raises an error

Keep in mind that this prevented pytest from even running a single test.

Also you can replace `typing.Pattern` with `typing.Callable` or `typing.Tuple` since those are also instances of the same class, which seems a bit more worrying.

And note that this applied to `M[int]()`, not `M[int]`.

**Comment by dmontagu** [comment]
> Keep in mind that this prevented pytest from even running a single test.

Fair point, that's a good reason to require that to be okay.

> If something has to wait until v3, then I sort of expect that it would have a special 'v3' label meaning "we're still recovering from releasing v2 and don't want too spend much energy thinking about this yet". What's the general attitude on your side?
>
> Would changing assert isinstance(M[int](), M[int]) be acceptable even in v3? It seems like a hell of a change.
> 
> (of course all this is only relevant if similar problems appear for types.GenericAlias)

I'm assuming they will occur for `types.GenericAlias` as well. The fundamental problem is still there — a parametrized generic generally refers to something that isn't a type, so subclass checks don't make sense. In the case of pydantic, it's an implementation detail that we return a subclass. I believe I have made the comment in other issues and possibly (hopefully) even the docs that you should avoid using a parametrized generic as the second argument to a subclass check, specifically because it's incompatible with standard library generics usage.

I agree it's a significant change, but I think the significantly improved consistency with standard library generics is worth breaking that if necessary.

I won't be surprised if we can figure something funny out that allows us to use a custom `__instancecheck__` or similar to keep the behavior consistent with current behavior. But if we can't, it sounds like you are inclined to agree it's significant enough to merit waiting until v3 if that ends up being necessary.

I think v3 is being considered more seriously than you might expect at this stage. The goal is not for it to be _nearly_ as big/challenging of a migration as v2, but to be a point where we can add some breaking changes (ideally minor, but still technically breaking) based on learnings from feedback on usage from users of v2.

At any rate, I haven't thought super deeply about this yet, at least in the context of this particular issue. It might make sense to give it a v3 label. If we could find a way to preserve all or nearly all behavior through smart use of dunder methods though I don't necessarily think it would need to wait.

**Comment by alexmojaki** [comment]
> a parametrized generic generally refers to something that isn't a type, so subclass checks don't make sense. In the case of pydantic, it's an implementation detail that we return a subclass.

I didn't realize it was an implementation detail, e.g. `test_generic_subclass` has `assert issubclass(B[int], B)`.

So would you say that changing `assert isinstance(M[int], type)` is acceptable and may or may not require v3?

**Comment by dmontagu** [comment]
I mean it's probably worth thinking about carefully, I would say I'm not convinced either way today, but I also know I haven't thought hard enough about it to have a confident decision. In particular, if there's broad consensus from the existing users of generic models about how this should be done, that would hold a lot of sway. My hope would be that most people are not relying on this sort of implementation-detail-level behavior though. (I mean, in particular, on the differences between pydantic generics and typing generics.)

If that's wrong and there are common/important/unfixable use cases, I'd be forced to reconsider.
